### PR TITLE
Switch from `.in` files to `[dependency-groups]`

### PR DESCRIPTION
**Pull request by sirosen**
`pip-tools` does not yet support Dependency Groups, but it is
in their tracker. Therefore, a small amount of shimming is needed to
apply it here.

1. Move the dependency groups from `requirements/*.in` files to
   `[dependency-groups]` in pyproject.toml
2. Add a tox environment (`freezedeps-print`) which installs
   `dependency-groups` (the package) and uses it to print dependency
   groups to new `.<name>.in` files, hidden and gitignored
3. Feed those files into `pip-compile` as before

Finally, run `tox r -m freezedeps` to confirm that the result works.

----
📚 Documentation preview 📚: https://globus-sdk-python--1089.org.readthedocs.build/en/1089/

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,41 @@
...
+    "typing-extensions>=4.0",
+    # although 'responses' is provided by the test requirements, it also
+    # must be installed for proper type-checking on globus_sdk._testing
+    "responses",
+]
+typing-mindeps = [
+    {include-group = "typing"},
+    "typing-extensions==4.0",
+]
+
```
Comment (sirosen) [inline pyproject.toml:38]: I just looked at this in the GitHub render. Should this table come after the `[project]` and `[tools.setuptools]` tables? Maybe push to the bottom of the file?

Comment (kurtmckee) [reply pyproject.toml:38]: I'm inclined to have this come after -- at least -- `[build-system]`. It's not tool-specific, so it seems fine to have it above the tool-specific sections. :+1:

File: requirements/.gitignore
Comment (kurtmckee) [inline requirements/.gitignore:1]: I'd prefer to have this in the top level `.gitignore`.

**Review by kurtmckee — APPROVED** [review]

File: requirements/.gitignore
Comment (sirosen) [reply requirements/.gitignore:1]: Sure! But I still want it restricted to this dir.

**Review by kurtmckee — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (sirosen) [reply pyproject.toml]: I put it after the `[project]` table. I'm still fussing about "is it right?! is it right?!" internally, but IMO we can live with this and move it around later if we dislike it.

**Pull request merged**